### PR TITLE
LCAM 945

### PR DESCRIPTION
**Pull request by skasthuri36326**
## What

[LCAM-945](https://dsdmoj.atlassian.net/browse/LCAM-945)

Describe what you did and why.
- Migrated the business logic from hardship.check_hardship  into the HardshipValidationService
- Added unit tests

## Checklist

Before you ask people to review this PR:

- [ ] Tests should be passing: `./gradlew test`
- [ ] Github should not be reporting conflicts; you should have recently run `git rebase main`.
- [ ] Avoid mixing whitespace changes with code changes in the same commit. These make diffs harder to read and conflicts more likely.
- [ ] You should have looked at the diff against main and ensured that nothing unexpected is included in your changes.
- [ ] You should have checked that the commit messages say why the change was made.

[LCAM-945]: https://dsdmoj.atlassian.net/browse/LCAM-945?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: crime-hardship/src/main/java/uk/gov/justice/laa/crime/hardship/service/HardshipValidationService.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+@Service
+@RequiredArgsConstructor
+public class HardshipValidationService {
+
+    public static final String HARDSHIP_REVIEW_STATUS_VALIDATION_MESSAGE = "Review Date must be entered for completed hardship";
+    public static final String NEW_WORK_REASON_VALIDATION_MESSAGE = "Review Reason must be entered for hardship";
+    public static final String SOLICITOR_DETAILS_VALIDATION_MESSAGE = "Solicitor Number of Hours must be entered when Solicitor Hourly Rate is specified";
+    public static final String FUNDING_SOURCES_VALIDATION_MESSAGE = "Amount and Date Expected must be entered for each detail in section ";
+    public static final String DENIEW_INCOME_VALIDATION_MESSAGE = "Amount, Frequency, and Reason must be entered for each detail in section ";
+    public static final String EXPENDITURE_VALIDATION_MESSAGE = "Amount, Frequency, and Reason must be entered for each detail in section ";
```
Comment (venkat980) [inline crime-hardship/src/main/java/uk/gov/justice/laa/crime/hardship/service/HardshipValidationService.java:28]: Can we avoid creating two constants for the same message? - Amount, Frequency, and Reason must be entered for each detail in section

File: crime-hardship/src/main/java/uk/gov/justice/laa/crime/hardship/service/HardshipValidationService.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+            if (fundingDescriptionWithoutFundingAmountOrDueDate(funding))
+                throw new ValidationException(FUNDING_SOURCES_VALIDATION_MESSAGE +funding.getDescription());
+        });
+    }
+
+    private void validateDeniedIncome(ApiPerformHardshipRequest apiPerformHardshipRequest) {
+        List<DeniedIncome> deniedIncomes = apiPerformHardshipRequest.getHardship().getDeniedIncome();
+        Optional.ofNullable(deniedIncomes).orElse(List.of()).forEach(deniedIncome -> {
+            if(deniedIncomeWithoutAmountOrFrequencyOrReasonNote(deniedIncome))
+                throw new ValidationException(DENIEW_INCOME_VALIDATION_MESSAGE +deniedIncome.getItemCode().getDescription());
```
Comment (venkat980) [inline crime-hardship/src/main/java/uk/gov/justice/laa/crime/hardship/service/HardshipValidationService.java:78]: Trivial - I see space missing after the + sign, maybe run the Code formatter before committing the changes?

File: crime-hardship/src/main/java/uk/gov/justice/laa/crime/hardship/service/HardshipValidationService.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        validateHardshipReviewNewWorkReason(apiPerformHardshipRequest);
+        validateSolicitorDetails(apiPerformHardshipRequest);
+        validateFundingSources(apiPerformHardshipRequest);
+        validateDeniedIncome(apiPerformHardshipRequest);
+        validateExpenditure(apiPerformHardshipRequest);
+        validateProgressionItems(apiPerformHardshipRequest);
+    }
+
+    private void validateHardshipReviewStatus(ApiPerformHardshipRequest apiPerformHardshipRequest) {
+        if(hardshipStatusIsCompleteWithoutReviewDate(apiPerformHardshipRequest)){
```
Comment (venkat980) [inline crime-hardship/src/main/java/uk/gov/justice/laa/crime/hardship/service/HardshipValidationService.java:42]: Code formatting would be nice here.

File: crime-hardship/src/main/java/uk/gov/justice/laa/crime/hardship/service/HardshipValidationService.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+                        isNull(expenditure.getReasonCode())));
+    }
+
+    private static boolean hardshipStatusIsCompleteWithoutReviewDate(ApiPerformHardshipRequest apiPerformHardshipRequest) {
+        return ((apiPerformHardshipRequest.getHardshipMetadata().getReviewStatus() == HardshipReviewStatus.COMPLETE)
+                && isNull(apiPerformHardshipRequest.getHardship().getReviewDate()));
+    }
+
+    private static boolean solicitorRateSpecifiedWithoutSolicitorHours(BigDecimal solicitorRate, Integer solicitorHours) {
+        return (solicitorRate.compareTo(BigDecimal.ZERO) > 0) && (solicitorHours == 0);
```
Comment (venkat980) [inline crime-hardship/src/main/java/uk/gov/justice/laa/crime/hardship/service/HardshipValidationService.java:118]: Please check the Integer comparison.

**Review by venkat980 — CHANGES_REQUESTED** [review]

File: crime-hardship/src/main/java/uk/gov/justice/laa/crime/hardship/service/HardshipValidationService.java (hunk 4 above)
Comment (skasthuri36326) [reply crime-hardship/src/main/java/uk/gov/justice/laa/crime/hardship/service/HardshipValidationService.java]: It should automatically unbox when compared to primitive. But added solicitorHours.intValue() to explicitly convert to primitive.

**Review by venkat980 — APPROVED** [review]

**Pull request merged**